Add SetToken function to NewAPIv4Client

atpons opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
Add function for setting Personal Access Token to `NewAPIv4Client`.
It is useful for call this package from other project. (e.g. Mattermost Bot Client)

e.g. 
```
cli := model.NewAPIv4Client("http://localhost:8065")
cli.SetToken("personal_access_token")
```

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-17471

mattermod commented:
Thanks @atpons for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

atpons commented:
Thanks for reviewing, @lieut-data. 

`MockSession` is used by `app/command_loadtest.go`,so  I think that the name `MockSession` is to be used for **mocking session**.

Functional pattern is idea for initializing Mattermost Client from other project e.g. Bot.

On the other hand, if not add this function, fixing name of `MockSession`, e.g. `SetToken` is an idea too.

lieut-data commented:
@atpons, agreed: `MockSession` is not a great name. I propose adding a /new/ `SetToken`, and changing the implementation of `MockSession` to use `SetToken`. We don't want to remove `MockSession`, as this would be a breaking API change.

aaronrothschild commented:
As a non-developer can someone explain what the purpose of this new call is?  Why are we adding it?

levb commented:
> As a non-developer can someone explain what the purpose of this new call is? Why are we adding it?

@aaronrothschild 2 different use-cases, the same code. @lieut-data and I are arguing we should not need a second copy, _maybe_ a better name, maybe not even that.

lindalumitchell commented:
To add a non-developer question, should this have a corresponding Jira ticket for tracking / testing?

And are any unit tests needed?

lieut-data commented:
@lindalumitchell, that's actually a good idea. There aren't any existing tests for `MockSession`, but I think it should be straightforward to add tests for both. @atpons, would you be open to helping us extend the unit test coverage here? I think it can just live in `model/client4_test.go` and trivially verify the assignment to `AuthToken` and `AuthType`.

I'll file a JIRA ticket to track.

atpons commented:
Thank you for comments.

`MockSession` is used by `app/command_loadtest.go`, and I changed to `SetToken` instead.

Because a narrow scope of this function, I thought it would not be a problem to export this function for OAuth and Personal Access Token. Also, the name of function is not strange, I think.

And, I added a test to verify that Bearer token is set correctly.

@lieut-data Thank you for your advice.

lieut-data commented on model/client4.go:
```diff
@@ -112,17 +112,11 @@ func BuildResponse(r *http.Response) *Response {
 	c.AuthType = HEADER_BEARER
 }
 
-func (c *Client4) MockSession(sessionToken string) {
```
Although /our/ code no longer uses `MockSession`, anyone importing `github.com/mattermost/mattermost-server/model` might still be using it, and removing a function amounts to semver breakage. Can we keep this function around, but just have it call the new `SetToken` instead? Might even be good to add a GoDoc:
```go
// MockSession is deprecated in favour of SetToken
func (c *Client4) MockSession(sessionToken string) {
```

atpons replied:
I didn't really understand semver. Fixed.

lieut-data commented on model/client4_test.go:
```diff
@@ -56,3 +56,14 @@ func TestClient4CreatePost(t *testing.T) {
+	client := NewAPIv4Client("http://example.com")
+	token := NewId()
+
+	client.AuthToken = token
```
Hmm, this doesn't actually test `SetToken`, as it does the work of that function itself. Can we change this to call `SetToken`, and then add a second unit test to verify the continued functionality of `MockSession`?

atpons replied:
Added a test to make sure MockSession works correctly. I tried to use the TestHelper in `model` package, but instead use httptest to check token because of import cycle.

lieut-data commented on model/client4_test.go:
```diff
@@ -56,3 +57,26 @@ func TestClient4CreatePost(t *testing.T) {
+	}))
+
+	client := NewAPIv4Client(server.URL)
+	client.AuthToken = expected
```
Thanks, @atpons -- we're still not actually testing `SetToken` here. Can you change this to read:
```go
	client := NewAPIv4Client(server.URL)
    client.SetToken(expected)
	_, resp := client.GetMe("")
	assert.Equal(t, http.StatusOK, resp.StatusCode)
```

ideally we'd also test `MockSession` for good measure:
```go
	client := NewAPIv4Client(server.URL)
    client.MockSession(expected)
	_, resp := client.GetMe("")
	assert.Equal(t, http.StatusOK, resp.StatusCode)
```

atpons replied:
I'm sorry. I will fix it now.
Do I need to call `MockSession` on `TestClient4SetToken` and specify **Deprecated** in the comment?
Or do I make 'TestClient4MockSession`?

atpons replied:
I think good to prepare 2 tests and if `MockSession` will be deprecated, to remove `TestClient4MockSession`.

hanzei commented:
@lindalumitchell Would you pleas check if this needs QA review?

@atpons Sorry for the deploy. Hoping to get this merged soon :tada:
